### PR TITLE
ISPN-12757 PartitionHandlingManagerImpl is unable to release locks

### PR DESCRIPTION
**Pull request by pruivo**
https://issues.redhat.com/browse/ISPN-12757

### CODE REVIEW

File: core/src/test/java/org/infinispan/partitionhandling/PessimisticTxPartitionHandlingReleaseLockTest.java
Comment (danberindei) [inline core/src/test/java/org/infinispan/partitionhandling/PessimisticTxPartitionHandlingReleaseLockTest.java:31]: It would be nice to also have a high-level description of the test in the javadoc

File: core/src/test/java/org/infinispan/partitionhandling/PessimisticTxPartitionHandlingReleaseLockTest.java
Comment (danberindei) [inline core/src/test/java/org/infinispan/partitionhandling/PessimisticTxPartitionHandlingReleaseLockTest.java:70]: Should partition handling be involved at all, if `cache0` is in a majority partition?

Also, what's blocking `cache0` from sending the commit command before the main thread installs the DISCARD protocol on manager1?

IMO you need more steps here to clarify what you expect to happen. E.g. wait for a particular cache topology in `cache0`, and verify that the transaction is stored in `PartitionHandlingManagerImpl.partialTransactions`, if that's what you mean with the comment.

**Review by danberindei — CHANGES_REQUESTED** [review]

File: core/src/test/java/org/infinispan/lock/ExplicitUnlockTest.java
Comment (pruivo) [reply core/src/test/java/org/infinispan/lock/ExplicitUnlockTest.java:141]: well, I would rather remove the test the change it :) 
we don't support users retrieving the `LockManager`, unlock keys and expect the transaction to commit fine...

File: core/src/test/java/org/infinispan/partitionhandling/PessimisticTxPartitionHandlingReleaseLockTest.java
Comment (pruivo) [reply core/src/test/java/org/infinispan/partitionhandling/PessimisticTxPartitionHandlingReleaseLockTest.java:70]: > Should partition handling be involved at all, if cache0 is in a majority partition?

to be on the safer side, yes. The `MultiTargetRequest` is updated in the JGroups view change event but the availability mode is only updated after the coordinator computes the partition. So, I don't think we can make a decision. Just give the transaction to the PartitionHandlingManager and let it decide when the stable topology is installed.

> Also, what's blocking cache0 from sending the commit command before the main thread installs the DISCARD protocol on manager1?

there is no commit and the prepare is blocked by `ControlledInboundHandler`

> IMO you need more steps here to clarify what you expect to happen.

hmmm ok. I didn't want to add complexity to the test.

**Comment by pruivo** [comment]
updated

**Comment by ryanemerson** [comment]
@danberindei Can you check the changes?

**Review by wburns — APPROVED** [review]
LGTM, will integrate soon unless @danberindei adds another comment :)

**Comment by wburns** [comment]
Integrated into master, thanks @pruivo !

Any additional changes can be done in a different PR.

File: core/src/test/java/org/infinispan/lock/ExplicitUnlockTest.java
Comment (danberindei) [reply core/src/test/java/org/infinispan/lock/ExplicitUnlockTest.java:141]: Why you say that @pruivo?

Even though `LockManager` is not `@public`, it is accessible via the public `AdvancedCache.getLockManager()` method, and there is no warning on the `unlock()` method that it is not supported.

File: core/src/test/java/org/infinispan/partitionhandling/PessimisticTxPartitionHandlingReleaseLockTest.java
Comment (danberindei) [reply core/src/test/java/org/infinispan/partitionhandling/PessimisticTxPartitionHandlingReleaseLockTest.java]: > to be on the safer side, yes. The MultiTargetRequest is updated in the JGroups view change event but the availability mode is only updated after the coordinator computes the partition. So, I don't think we can make a decision. Just give the transaction to the PartitionHandlingManager and let it decide when the stable topology is installed.

Does this mean `StateTransferInterceptor` no longer retries commit/1PC prepare commands?
I see the code is still there, but maybe it only runs with partition handling disabled.

> hmmm ok. I didn't want to add complexity to the test.

From my point of view it is *simplifying* the reading of the test :)

BTW, I find the new comment a bit confusing as well

```
//after the view change, the transaction is handled by the PartitionHandlingManager which finishes the commit
```

After reading the whole method a couple of times I understood that the transaction is passed to PartitionHandlingManager and the commit call returns, but the first time I thought it meant "wait until PartitionHandlingManager retries the commit and the transaction is completed"

File: core/src/test/java/org/infinispan/lock/ExplicitUnlockTest.java
Comment (pruivo) [reply core/src/test/java/org/infinispan/lock/ExplicitUnlockTest.java:141]: well, I don't see any use case where a user needs to acquire/release locks from the `LockManager`. Even if there is a weird use case, I'm pretty sure data consistency is automatically lost. 
Since we have clustered-locks available, I think we could remove the method `AdvancedCache.getLockManager()`

File: core/src/test/java/org/infinispan/partitionhandling/PessimisticTxPartitionHandlingReleaseLockTest.java
Comment (pruivo) [reply core/src/test/java/org/infinispan/partitionhandling/PessimisticTxPartitionHandlingReleaseLockTest.java]: > Does this mean `StateTransferInterceptor` no longer retries commit/1PC prepare commands?
> I see the code is still there, but maybe it only runs with partition handling disabled.

No, Partition handling only handles a single case: a crash from a node involved in the transaction. All other cases (leaves/join/crashes from other nodes) are retried.

> After reading the whole method a couple of times I understood that the transaction is passed to PartitionHandlingManager and the commit call returns, but the first time I thought it meant "wait until PartitionHandlingManager retries the commit and the transaction is completed"

Sorry about that :)

File: core/src/test/java/org/infinispan/lock/ExplicitUnlockTest.java
Comment (danberindei) [reply core/src/test/java/org/infinispan/lock/ExplicitUnlockTest.java:141]: I agree `LockManager` is not a good API, but we need to check if some stuff like `getOwner(key)` or `getNumberOfLocksHeld()` should be kept in the API for diagnostics.

File: core/src/test/java/org/infinispan/partitionhandling/PessimisticTxPartitionHandlingReleaseLockTest.java
Comment (danberindei) [reply core/src/test/java/org/infinispan/partitionhandling/PessimisticTxPartitionHandlingReleaseLockTest.java]: > No, Partition handling only handles a single case: a crash from a node involved in the transaction. All other cases (leaves/join/crashes from other nodes) are retried.

I see, I mistook your "Just give the transaction to the PartitionHandlingManager" comment to mean that PHM will always take care of completing the transaction after a `CacheNotFoundResponse`.

**Pull request merged**
